CC-3950: Change the HDFS sink task to return to Connect offsets for records committed to HDFS

rhauch opened the pull request:
Modified the task’s `preCommit(…)` method to always return the offsets for only those records that have been committed to HDFS. Connect will then use the consumer to commit those offsets to the `__consumer_offsets` topic, which allows for simple monitoring of the approximate lag of the connector.

First, this change does not affect the HDFS connector’s exactly once behavior, because the connector still writes and uses the offsets it records in the files in HDFS; as long as these files are found in HDFS for the topic partitions, the HDFS connector will use these to determine where Connect should begin consuming. If the HDFS connector does not find the files in HDFS, it will let Connect and the consumer determine where to start, which will be based upon the committed offsets if found or upon the `consumer.auto.offset.reset` value specified in the worker config (which for Connect defaults to `earliest`) if no offsets are found. None of this behavior has changed.

What has changed is that, starting with CP 4.0.1, the HDFS sink connector never returned offsets, which means Connect never updating the offsets in `__consumer_offsets` for the topic partitions the connector read. Users could still change these offsets, and the connector would use these if no offsets were found in HDFS, but it never updated these offsets.

Now, the HDFS connector does return the offsets of the records from the various topic partitions that were most recently committed to HDFS, and Connect will thus periodically (based upon `consumer.offset.flush.interval.ms` in the worker config) write these offsets to the `__consumer_offsets` topic using the appropriate consumer group for the connector.

Because the offsets are committed to `__consumer_offsets` asynchronously, it is likely that the offsets in `__consumer_offsets` will be somewhat behind where the connector actually is. In fact, the offsets in `__consumer_offsets` should never be ahead of the records that the connector has written to HDFS. This allows the offsets as a simple way to estimate the upper limit of the actual connector lag.

rhauch commented on src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java:
```diff
@@ -480,8 +482,13 @@ public void buffer(SinkRecord sinkRecord) {
+   * @return the last committed offset; may be null
+   */
+  public Long committedOffset() {
+    return committedOffset;
```
The `offset` value was *almost* what we needed, but unfortunately it is modified with the first record the writer processes. This means that it does not always represent the last offset that was committed.

Rather than change the logic that relied upon this behavior, I chose to introduce a new `committedOffset` member to represent the last offset that was committed, and set it equal to `offset` whenever the writer commits its file.

rhauch replied:
Interesting. I tried changing it locally, and it felt awkward (e.g., in the JavaDoc for `offsetToCommit()` because, for the most part, this `TopicPartitionWriter` uses "commit" specifically for committing to HDFS. `offsetToCommit()` just felt like it was the offset that had yet to be written to HDFS.

So I guess I prefer leaving it `committedOffset`.

kkonstantine replied:
Not insisting here. Just wanted to make an observation. `Commit` is a very overloaded term in these connectors indeed.

rhauch commented on src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java:
```diff
@@ -558,6 +565,7 @@ private void readOffset() throws ConnectException {
     );
     if (fileStatusWithMaxOffset != null) {
       offset = FileUtils.extractOffset(fileStatusWithMaxOffset.getPath().getName()) + 1;
+      committedOffset = offset;
```
If we read the offset from HDFS files, we update the committed offset here.

wicknicks replied:
nit: the committedOffset is actually `offset - 1`. `offset` here is the value we will next read from.

kkonstantine replied:
(pressed resolved by accident, apologies). 
Setting 
`committedOffset = FileUtils.extractOffset(fileStatusWithMaxOffset.getPath().getName());`
and then
`offset = committedOffset + 1`
probably makes more sense (fyi `preCommit` won't accidentally commit offsets that have not been consumed, we might delay offset commit with a wrong setting).

rhauch replied:
I just made this same change re setting `committedOffset` and `offset`. Great minds... :-D

What do you mean by "we might delay offset commit with a wrong setting"?

kkonstantine replied:
I meant that there's no risk to commit offsets ahead of the current position with `preCommit`. 
See: https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L410-L421

It's just that the recommendation will be ignored and therefore the lag won't be represented as soon as it would be possible.

rhauch replied:
Gotcha. The WorkerSinkTask will prevent committing offsets for records that haven't yet been consumed, but IMO it'd be better to not rely upon that feature since it may result in a larger lag (as measured by the committed offset), especially if the connector always returns a yet-to-be-consumed offset.

I think we're all on the same page, and the updated code uses the correct offset anyway.

rhauch commented on src/main/java/io/confluent/connect/hdfs/DataWriter.java:
```diff
@@ -505,7 +505,10 @@ public Partitioner getPartitioner() {
+      Long committedOffset = topicPartitionWriters.get(tp).committedOffset();
+      if (committedOffset != null) {
+        offsets.put(tp, committedOffset);
+      }
```
We only include the offset for topic partitions with records that were committed by the TopicPartitionWriter instances to HDFS. If `committedOffset()` returns null, we don't even include it.

Despite the name, the previous incarnation of this method didn't return only the committed offsets; as mentioned in another comment, the offset of the first record each TopicPartitionWriter processed would be returned, whether or not this was actually written. And because this method was only used in tests, I felt it was acceptable to change the behavior to return only the committed offset.

wicknicks commented on src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java:
```diff
@@ -767,6 +775,7 @@ private void commitFile(String encodedPartition) {
     startOffsets.remove(encodedPartition);
     offsets.remove(encodedPartition);
     offset = offset + recordCounter;
+    committedOffset = offset;
```
same here? `committedOffset` is `offset - 1`.

rhauch replied:
I was thinking the same thing - `endOffset` seems to be more accurate.

kkonstantine replied:
I think the choice here is, do we want to report the most accurate number or the most safest number. (`offset-1` vs `endOffset`. I think that's what it is here, but I might have to look closer). Maybe go with the safest option of `endOffset`?

rhauch commented on src/main/java/io/confluent/connect/hdfs/HdfsSinkTask.java:
```diff
@@ -119,13 +119,19 @@ public void put(Collection<SinkRecord> records) throws ConnectException {
   ) {
-    // return an empty map so Connect doesn't commit offsets
-    return Collections.emptyMap();
+    // Although the connector manages offsets via the WAL, we still want to have Connect commit
```
ack. will reword.

rhauch commented on src/main/java/io/confluent/connect/hdfs/HdfsSinkTask.java:
```diff
@@ -119,13 +119,19 @@ public void put(Collection<SinkRecord> records) throws ConnectException {
+    // Although the connector manages offsets via the WAL, we still want to have Connect commit
+    // the consumer offsets for records this task has consumed from its topic partitions and
+    // committed to HDFS.
+    Map<TopicPartition, OffsetAndMetadata> result = new HashMap<>();
```
I tried this, but not sure what I was doing wrong. I agree, go streams or go home.

wicknicks replied:
I just realized, if we plan to backport this to `4.0.x`, then we can't use java 8 apis (unless we want to have a separate PR for `4.0.x`).

rhauch replied:
I'll submit the PR against `4.0.x` and change to imperative style, and create a separate PR that will switch to the functional style (which IMO is far better).

wicknicks replied:
Makes backporting on these lines a bit harder, otherwise should be alright.

rhauch commented:
@wicknicks , @kkonstantine: thanks for reviewing. I've added a commit that incorporates most of your suggestions (all but the renaming of `committedOffsets()` to `offsetsToCommit()` as described in the inline comments.

Would appreciate another review.

kkonstantine commented on src/test/java/io/confluent/connect/hdfs/avro/DataWriterAvroTest.java:
```diff
@@ -396,7 +396,7 @@ public void testFlushPartialFile() throws Exception {
     assertTrue(committedOffsets.containsKey(TOPIC_PARTITION));
     long previousOffset = committedOffsets.get(TOPIC_PARTITION);
-    assertEquals(NUMBER_OF_RECORDS, previousOffset);
+    assertEquals(NUMBER_OF_RECORDS-1, previousOffset);
```
nit: that tells me we don't run checkstyle on tests (?)

kkonstantine replied:
@rhauch this needs spaces around `-`

rhauch replied:
fixed

rhauch commented:
@kkonstantine , @wicknicks: I've rebased the PR on `4.0.x` and switched from functional to imperative code for `HdfsSinkTask.preCommit()`. Otherwise, this is the same set of code as previously reviewed. I've removed the "WIP" from the title and description, so this is ready for the final review.

rhauch commented:
@wicknicks, @kkonstantine: I made a small change to some of the log messages and the formatting of a test. I'll give you a chance to review and disapprove. :-D
